### PR TITLE
Support more escape codes inside labels 

### PR DESCRIPTION
**Pull request by sbstp**
Fix for #214 

Now escapes `|`, `&`, `<`, `>` and `\`. Turned out to be a bit more tricky than I anticipated, but it's all good now.

### CODE REVIEW

File: src/dot.rs [hunk 1]
```diff
@@ -171,10 +171,13 @@ impl<W> fmt::Write for Escaper<W>
 
     fn write_char(&mut self, c: char) -> fmt::Result {
         match c {
-            '"' => try!(self.0.write_char('\\')),
+            '"' | '\\' | '|' | '<' | '>' | '&' => self.0.write_char('\\')?,
             // \l is for left justified linebreak
-            '\n' => return self.0.write_str(r#"\l"#),
-            _   => { }
+            '\n' => {
+                self.0.write_char('\\')?;
```
Comment (Dylan-DPC) [inline src/dot.rs]: error: the `?` operator is not stable (see issue #31436)

File: src/dot.rs [hunk 2]
```diff
@@ -171,10 +171,13 @@ impl<W> fmt::Write for Escaper<W>
 
     fn write_char(&mut self, c: char) -> fmt::Result {
         match c {
-            '"' => try!(self.0.write_char('\\')),
+            '"' | '\\' | '|' | '<' | '>' | '&' => self.0.write_char('\\')?,
```
Comment (Dylan-DPC) [inline src/dot.rs]: error: the `?` operator is not stable (see issue #31436)

**Comment by Dylan-DPC** [comment]
Hey. Thanks for the PR

We need to support the stable rust version and thus have to depend on stable features as much as possible.

**Comment by sbstp** [comment]
`?` is stable since [1.13 (2016-11-10)](https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1130-2016-11-10),  but it wasn't in 1.12, so I've updated the PR match that. Everything should be fine now.

From reading the spec again, I don't think escaping `<`, `>` and `&` is required because HTML is not parsed inside strings delimited by quotes.

**Comment by Dylan-DPC** [comment]
Yeah that's what i meant. Maybe it is time to bump up the minimum rustc version soon.

File: src/dot.rs [hunk 3]
```diff
@@ -171,10 +171,13 @@ impl<W> fmt::Write for Escaper<W>
...
     fn write_char(&mut self, c: char) -> fmt::Result {
         match c {
-            '"' => try!(self.0.write_char('\\')),
+            '"' | '\\' | '|' | '<' | '>' | '&' => try!(self.0.write_char('\\')),
             // \l is for left justified linebreak
-            '\n' => return self.0.write_str(r#"\l"#),
-            _   => { }
+            '\n' => {
+                try!(self.0.write_char('\\'));
+                return self.0.write_char('l');
```
Comment (bluss) [inline src/dot.rs]: Why did you change from write_str to 2 write_chars here? I guess the difference is trivial, but normally write_str should be better w.r.t code quality.

File: src/dot.rs [hunk 4]
```diff
@@ -188,7 +191,7 @@ impl<T> fmt::Display for Escaped<T>
 {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         if f.alternate() {
-            write!(&mut Escaper(f), "{:#}\\l", &self.0)
+            write!(&mut Escaper(f), "{:#}\n", &self.0)
```
Comment (bluss) [inline src/dot.rs]: Is there a reason for this change?

**Comment by bluss** [comment]
Bumping the minimum rust version sounds good for the next version (and thus current master). Not something super new, but for example Rust 1.18 maybe. That can be a different PR.

Here, it sounds like we should adjust the escaping to match the spec and don't escape those extra chars then?

File: src/dot.rs (hunk 4 above)
Comment (sbstp) [reply src/dot.rs]: Yes, since backslashes are escaped now, `\\l` turns to `\\\\l` which is not a the intent. Since `\\l` turns to `\n`, I put the converted form there directly to avoid escaping `\\l`.

File: src/dot.rs (hunk 3 above)
Comment (sbstp) [reply src/dot.rs]: I had double escaping issues at some point because I was using `self.write_str`, but I think it might be fine if I use `self.0.write_str`.

File: src/dot.rs (hunk 4 above)
Comment (bluss) [reply src/dot.rs]: ah, of course. nice

**Comment by sbstp** [comment]
The spec is not very detailed, they don't talk about escaping `|` for instance, but it renders differently when escaped. Escaping the html `<`, `>` and `&` seems not to be necessary but it does not affect rendering since graphviz ignores lone backslashes. So I think I'll remove the html escape for keep the backslashes and the pipe symbol.

File: src/dot.rs (hunk 4 above)
Comment (sbstp) [reply src/dot.rs]: I got a little confused there. `\\l` would turn to `\\\\l`, and in order to get `\\l` in the output, `\n` is used.

**Comment by sbstp** [comment]
The pipe is relevant when a [node is in record mode](https://www.graphviz.org/doc/info/shapes.html#record). Since the Dot generator does not use record mode, I believe it's safe not to escape it.

**Comment by sbstp** [comment]
Mixing Debug with escaping can be surprisingly confusing. I left a comment for the next person maintaining this.

**Comment by sbstp** [comment]
The nightly build seems to have an issue which is not related to PR so I think this is finally ready.

**Comment by bluss** [comment]
Thanks for this!
